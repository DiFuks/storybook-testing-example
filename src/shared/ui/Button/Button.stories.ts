import type { Meta, StoryObj } from '@storybook/react';
import { expect, fn, userEvent, within } from '@storybook/test';

import { Button } from './Button';

const meta = {
	component: Button,
	// eslint-disable-next-line @typescript-eslint/quotes
	tags: ['autodocs'],
	args: { onClick: fn() },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
	args: {
		children: `Primary`,
		variant: `primary`,
	},
	play: async ({ canvasElement, args }) => {
		const canvas = within(canvasElement);

		await userEvent.click(canvas.getByRole(`button`));

		await expect(args.onClick).toHaveBeenCalled();
	},
};

export const Secondary: Story = {
	args: {
		children: `Secondary`,
		variant: `secondary`,
	},
};
