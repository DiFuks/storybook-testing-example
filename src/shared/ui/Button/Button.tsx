import { type FC, type HTMLAttributes, type MouseEventHandler, useState } from 'react';
import styled from 'styled-components';

type ButtonVariant = `primary` | `secondary`;

interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
	variant?: ButtonVariant;
}

export const Button: FC<ButtonProps> = ({ variant = `primary`, onClick, children, ...props }) => {
	const [isClicked, setIsClicked] = useState(false);

	const handleClick: MouseEventHandler<HTMLButtonElement> = event => {
		setIsClicked(prevIsClicked => !prevIsClicked);
		onClick?.(event);
	};

	return (
		<ButtonStyled {...props} variant={variant} onClick={handleClick}>
			{isClicked ? `Clicked!` : children}
		</ButtonStyled>
	);
};

interface ButtonStyledProps {
	variant: ButtonVariant;
}

const ButtonStyled = styled.button<ButtonStyledProps>`
	padding: 10px 20px;
	border: none;
	border-radius: 5px;
	color: white;
	background-color: ${props => (props.variant === `primary` ? `#007bff` : `#6c757d`)};
	cursor: pointer;
	font-size: 16px;
	transition: background-color 0.2s;

	&:hover {
		background-color: ${props => (props.variant === `primary` ? `#0056b3` : `#545b62`)};
	}

	&:focus {
		outline: none;
		box-shadow: 0 0 0 2px rgba(0, 123, 255, 0.25);
	}
`;
