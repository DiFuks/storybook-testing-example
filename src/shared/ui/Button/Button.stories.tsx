import { type FC, type MouseEventHandler, useState } from 'react';
import type { Meta, StoryObj } from '@storybook/react';
import { expect, fn, userEvent, within } from '@storybook/test';

import { Button, type ButtonProps } from './Button';

const ButtonWithHooks: FC<ButtonProps> = ({ onClick, children, ...props }) => {
	const [isClicked, setIsClicked] = useState(false);

	const handleClick: MouseEventHandler<HTMLButtonElement> = event => {
		setIsClicked(prevIsClicked => !prevIsClicked);
		onClick?.(event);
	};

	return (
		<Button {...props} onClick={handleClick}>
			{isClicked ? `Clicked!` : children}
		</Button>
	);
};

const meta = {
	component: Button,
	// eslint-disable-next-line @typescript-eslint/quotes
	tags: ['autodocs'],
	args: { onClick: fn() },
	render: args => <ButtonWithHooks {...args} />,
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
	args: {
		children: `Primary`,
		variant: `primary`,
	},
	play: async ({ canvasElement, args, step }) => {
		const canvas = within(canvasElement);
		const button = canvas.getByRole(`button`);

		await expect(button).toHaveTextContent(/Primary/);

		await step(`First click`, async () => {
			await userEvent.click(button);

			await expect(args.onClick).toHaveBeenCalled();
			await expect(button).toHaveTextContent(/Clicked!/);
		});

		await step(`Second click`, async () => {
			await userEvent.click(button);

			await expect(button).toHaveTextContent(/Primary/);
		});
	},
};

export const Secondary: Story = {
	args: {
		children: `Secondary`,
		variant: `secondary`,
	},
};
