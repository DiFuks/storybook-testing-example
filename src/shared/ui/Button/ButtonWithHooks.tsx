import { type FC, type MouseEventHandler, useState } from 'react';

import { Button, type ButtonProps } from './Button';

export const ButtonWithHooks: FC<ButtonProps> = ({ onClick, children, ...props }) => {
	const [isClicked, setIsClicked] = useState(false);

	const handleClick: MouseEventHandler<HTMLButtonElement> = event => {
		setIsClicked(prevIsClicked => !prevIsClicked);
		onClick?.(event);
	};

	return (
		<Button {...props} onClick={handleClick}>
			{isClicked ? `Clicked!` : children}
		</Button>
	);
};
