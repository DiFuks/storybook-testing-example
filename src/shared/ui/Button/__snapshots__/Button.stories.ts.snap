// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Primary play-test 1`] = `
<button variant="primary"
        class="Button__ButtonStyled-sc-4dd5c220-0 eZeqKy"
>
  Primary
</button>
`;

exports[`Button Secondary smoke-test 1`] = `
<button variant="secondary"
        class="Button__ButtonStyled-sc-4dd5c220-0 xznk"
>
  Secondary
</button>
`;
